pennylane/templates/subroutines/qrom.py

```diff
@@ -0,0 +1,300 @@
+        control_wires (Sequence[int]): the wires where the indexes are specified
+        target_wires (Sequence[int]): the wires where the bitstring is loaded
+        work_wires (Sequence[int]): the auxiliary wires used for the computation
+        clean (bool): if True, the work wires are not altered by operator, default is ``True``
+
+    **Example**
+
+    In this example the QROM is applied and the target wires are measured to get the third bitstring.
+
+    .. code-block::
+
+        bitstrings = ["010", "111", "110", "000"]
+
+        dev = qml.device("default.qubit", shots = 1)
+
+        @qml.qnode(dev)
+        def circuit():
+
+            # third index
+            qml.BasisEmbedding(2, wires = [0,1])
+
+            qml.QROM(bitstrings = bitstrings,
+                    control_wires = [0,1],
+                    target_wires = [2,3,4],
+                    work_wires = [5,6,7])
+
+          return qml.sample(wires = [2,3,4])
+
+    .. code-block:: pycon
+
+        >>> print(circuit())
+        [1 1 0]
+
+
+    .. details::
+        :title: Usage Details
+
+        This template takes as input three different sets of wires. The first one in ``control_wires``. This register
+        makes reference to the wires where we will introduce the index. Therefore, if we have :math:`m` bitstrings, we need
+        at least :math:`\lceil \log_2(m)\rceil` wires.
+
+        The second set of wires is ``target_wires``. These are the wires where the bitstrings will be stored.
+        For instance, if the bitstring is "0110", we will need four target wires. Internally the bitstrings are
+        encoded using the :class:`~.BasisEmbedding` template.
+
+
+        The ``work_wires`` are the auxiliary qubits used by the template to reduce the number of gates required.
+        Let :math:`k` be the number of work wires. If :math:`k = 0`, the template is equivalent to executing :class:`~.Select`.
+        Following the idea in `arXiv:1812.00954 <https://arxiv.org/abs/1812.00954>`__, auxiliary qubits can be used to
+        load in parallel more than one bit string. Let :math:`\lambda` be
+        the number of bitstrings we want to store in parallel, which it is assumed to be a power of :math:`2`.
+        Then, :math:`k = l \cdot (\lambda-1)` work wires are needed,
+        where :math:`l` is the length of the bitstrings.
+
+        The QROM template has two variants. The first one (``clean = False``) is based on the previous paper that alterates the state in the ``work_wires``.
+        The second one (``clean = True``), based on `arXiv:1902.02134 <https://arxiv.org/abs/1902.02134>`__, solves that issue by
+        returning ``work_wires`` to their initial state. This technique is able to work with ``work_wires`` that are not
+        initialized to zero.
+
+    """
+
+    def __init__(
+        self, bitstrings, control_wires, target_wires, work_wires, clean=True, id=None
+    ):  # pylint: disable=too-many-arguments
+
```

Comment:
Will be good to have error messages when wrong number of wires are entered. For example, if the number of target wires is not equal to the length of the bitstring. Or when the number of control wires is not enough to encode the index.

Reply:
^^ +1, especially, it would be great if there was a function which would tell you how many wires you need (work, control and target) for a given set of bit strings 👍🏼